docs/reference/providers/custom/building-custom-provider.md

```diff
@@ -0,0 +1,263 @@
+                "type": 2,
+                "Config": null
+              }
+            ]
+          }
+        }
+      },
+      {
+        "name": "hello",
+        "type": 1,
+        "Config": {
+          "ExecutionUnit": {
+            "Source": {
+              "Image": {
+                "uri": "custom-provider-demonstration-hello"
+              }
+            },
+            "workers": 1,
+            "type": "default"
+          }
+        }
+      }
+    ]
+  },
+  "attributes": {
+    "project": "custom-provider-demonstration",
+    "stack": "custom-stack"
+  }
+}
+```
+
+The spec contains a list of resources and global attributes that can be passed to the deployment provider. The list of resources for this deployment contains an API, bucket, and execution unit with the information required to configure them for the cloud. The default project and stack attributes are also provided. You will notice in the execution unit config the image is provided with a URI pointing to the local image name.
+
+### Setting up the server
+
+The deployment provider requires an implementation for deploying and a method for destroying the cloud resources, called `Up` and `Down` respectively.
+
+```golang
+type DeployServer struct {
+}
+
+func (d *DeployServer) Up(request *deploy.DeployUpRequest, stream deploy.DeployService_UpServer) error {
+}
+
+func (d *DeployServer) Down(request *deploy.DeployDownRequest, stream deploy.DeployService_DownServer) error {
+}
+```
+
+The `request` contains the resource specification, and the `stream` is to return logs to the CLI for output.
+
+Using this base, the provider can be registered with the CLI. The following provides a utility function for starting a gRPC server and registering the deployment provider with the CLI. The package `v1` is the compiled protocol buffers.
+
+```golang
+import (
+  ...
+  v1 "github.com/nitrictech/nitric/core/pkg/api/nitric/deploy/v1"
+	"github.com/nitrictech/nitric/core/pkg/utils"
+	"google.golang.org/grpc"
+  ...
+)
+
+func StartServer(deploySrv v1.DeployServiceServer) {
+  // Check the port is available over TCP
+	lis, err := net.Listen("tcp", ":50051"))
+	if err != nil {
+		log.Fatalf("error listening on port %s %v", port, err)
+	}
+
+  // Create the gRPC server
+	srv := grpc.NewServer()
+
+  // Register the deployment server address with the CLI
+	v1.RegisterDeployServiceServer(srv, deploySrv)
+
+	fmt.Printf("Deployment server started on %s\n", lis.Addr().String())
+
+  // Start the gRPC server
+	err = srv.Serve(lis)
+	if err != nil {
+		log.Fatalf("error serving requests %v", err)
+	}
+}
+```
+
+> As protocol buffers can be compiled to any language, the deployment providers can be written in any language.
+
+In the future, we may release developer resources that make writing the custom providers easier for the languages that Nitric supports. This would remove the need for this boilerplate so that you can focus on writing your code. This may look like the following:
```

Comment:
Feels like this section needs a heading and a link where reader can contribute feedback on this approach. It would be great to know if people want us to do this